consensus/hotstuff/cruisectl/proposale_timing.go

```diff
@@ -0,0 +1,151 @@
+// at time `t := TimedBlock.TimeObserved` and applying the limits of authority yields `d := ConstrainedBlockTime()`
+// Then, `TargetPublicationTime(..)` returns `t + d` as the target publication time for the child block.
+//
+// By convention, happyPathBlockTime should be treated as immutable.
+// TODO: any additional logic for assiting the EventHandler in determining the applied delay should be added to the ControllerViewDuration
+type happyPathBlockTime struct {
+	TimedBlock    // latest block observed by the controller, including the time stamp when the controller received the block [UTC]
+	*TimingConfig // timing configuration for the controller, for retrieving the controller's limits of authority
+
+	// unconstrainedBlockTime is the delay, relative to `TimedBlock.TimeObserved` when the controller would
+	// like the child block to be published. Caution, no limits of authority have been applied to this value yet.
+	// The final controller output after applying the limits of authority is returned by function `ConstrainedBlockTime`
+	unconstrainedBlockTime time.Duration // desired duration until releasing the child block, measured from `TimedBlock.TimeObserved`
+
+	constrainedBlockTime time.Duration // block time _after_ applying limits of authority to unconstrainedBlockTime
+}
+
+var _ ProposalTiming = (*happyPathBlockTime)(nil)
+
+// newHappyPathBlockTime instantiates a new happyPathBlockTime
+func newHappyPathBlockTime(timedBlock TimedBlock, unconstrainedBlockTime time.Duration, timingConfig *TimingConfig) *happyPathBlockTime {
+	return &happyPathBlockTime{
+		TimingConfig:           timingConfig,
+		TimedBlock:             timedBlock,
+		unconstrainedBlockTime: unconstrainedBlockTime,
+		constrainedBlockTime:   min(max(unconstrainedBlockTime, timingConfig.MinProposalDuration), timingConfig.MaxProposalDuration),
+	}
+}
+
+func (pt *happyPathBlockTime) ObservationView() uint64             { return pt.Block.View }
+func (pt *happyPathBlockTime) ObservationTime() time.Time          { return pt.TimeObserved }
+func (pt *happyPathBlockTime) ConstrainedBlockTime() time.Duration { return pt.constrainedBlockTime }
+
+// TargetPublicationTime operates in two possible modes:
+//  1. If `parentBlockId` matches our `TimedBlock`, i.e. the EventHandler is just building the child block, then
+//     we return `TimedBlock.TimeObserved + ConstrainedBlockTime` as the target publication time for the child block.
+//  2. If `parentBlockId` does _not_ match our `TimedBlock`, the EventHandler should release the block immediately.
+//     This heuristic is based on the intuition that Block time is expected to be very long when deviating from the happy path.
+func (pt *happyPathBlockTime) TargetPublicationTime(proposalView uint64, timeViewEntered time.Time, parentBlockId flow.Identifier) time.Time {
+	if parentBlockId != pt.Block.BlockID {
+		return timeViewEntered // broadcast immediately
+	}
+	return pt.TimeObserved.Add(pt.ConstrainedBlockTime()) // happy path
```

Comment:
Agree that we should be thinking about it. 

Unfortunately, I think its not quite as simple. I am worried that your solution might yield an output outside of the limits of authority. And potentially, it might add more error. In my mind, this is the break-down:
* dominant first order behaviour:  happy path (roughly 95% of views) 
* second order system behaviour: a few consecutive primaries (maybe 2 - 5, or 6) are offline
* third order and higher: participants recovering from crashes, consensus nodes being misconfigured and producing erratic block timings, networking and connection issues, ...

As we have the first order covered by the controller, we are discussing strategies for handling second order effects.

Reply:
The more I am thinking about it, the more I tend towards the following
```golang
func (pt *happyPathBlockTime) TargetPublicationTime(proposalView uint64, timeViewEntered time.Time, parentBlockId flow.Identifier) time.Time {
	if proposalView < pt.Block.View { // views should be monotonically increasing, this is a fallback for numeric stability
		return timeViewEntered // broadcast immediately
	}
	d := pt.ConstrainedBlockTime() * time.Duration(proposalView-pt.Block.View)
	return pt.TimeObserved.Add(d) // happy path
}
```